### PR TITLE
Job completion check based on container status.

### PR DESCRIPTION
**Pull request by shrinandj**
This change modifies the way a Metaflow step running as a Kubernetes job is marked completed. Previously, Metaflow would wait until the job itself is marked as completed. However, this can take much longer than the time the container running in the job is marked completed.

There don't seem to be any obvious tunables in K8s for making it faster to mark a job as completed.

Given that a Metaflow job runs a single container, this change relies on checking that containers' status.

Testing Done:

- Basic linear flows run to completion.
- Foreach based flows run to completion.
- Flows that have failed steps show up as failed. Logs were shown on stderr and the metaflow ui.
  - Tested with a step that explicitly did a sys.exit()
  - Tested with a step that had it's pod explicitly terminated (kubectl delete pod)
- Terminated one step in a foreach. Verified that the run shows up as failed
- Tested that flows run just fine with Argo workflows

### CODE REVIEW

File: metaflow/plugins/kubernetes/kubernetes_job.py [hunk 1]
```diff
@@ -339,15 +339,16 @@ def __init__(self, client, name, uid, namespace):
         self._job = self._fetch_job()
         self._pod = self._fetch_pod()
 
-        import atexit
+        import signal
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_job.py:342]: what's the motivation for this change?

File: metaflow/plugins/kubernetes/kubernetes_job.py [hunk 2]
```diff
@@ -563,9 +564,24 @@ def has_failed(self):
...
+        if self.is_done:
+            return False
+
+        container_statuses = self._pod.get("status", {}).get("container_statuses", [])
+        if not container_statuses:
+            # Yet to start, most likely
+            return True
+        if len(container_statuses) < 1:
+            # Yet to start, most likely
+            return True
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_job.py:576]: should we return from pod phase here and above?

File: metaflow/plugins/kubernetes/kubernetes_job.py [hunk 3]
```diff
@@ -563,9 +564,24 @@ def has_failed(self):
...
+        # There is guaranteed to be only 1 container
+        cstatus = container_statuses[0]
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_job.py:578]: can we fetch the container status using the name of the container? that will guard against the scenario if/when we have more than one container.

**Review by savingoyal — CHANGES_REQUESTED** [review]

File: metaflow/plugins/kubernetes/kubernetes_job.py (hunk 1 above)
Comment (shrinandj) [reply metaflow/plugins/kubernetes/kubernetes_job.py:342]: Without this change, the code would try to fetch the job object and then patch it to set parallelism to 0. This even when knowing that the container has actually completed execution. So, it's only a matter of time when the job is also marked completed. Having the original code using `atexit` as is reduces the latency improvements introduced in this change.

AFAICT, the original goal of this code was to cleanup the pods (best-effort) in the event of the user-created process running Metaflow is terminated. The change continues to do that.

Let me know if there was any additional motivation for having this and I can test that scenario out.

File: metaflow/plugins/kubernetes/kubernetes_job.py (hunk 2 above)
Comment (shrinandj) [reply metaflow/plugins/kubernetes/kubernetes_job.py:576]: Done

File: metaflow/plugins/kubernetes/kubernetes_job.py (hunk 3 above)
Comment (shrinandj) [reply metaflow/plugins/kubernetes/kubernetes_job.py:578]: name of the container is derived from the step name and is not available in the `RunningJob`. 

I'll change the code to iterate over all containers and if any one of them is NOT terminated, return True for `is_running`. That should protect against the case where there are multiple containers.

File: metaflow/plugins/kubernetes/kubernetes_job.py [hunk 4]
```diff
@@ -563,9 +564,29 @@ def has_failed(self):
     @property
     def is_running(self):
         # Returns true if the pod is running.
-        return not self.is_done and (
-            self._pod.get("status", {}).get("phase") == "Running"
-        )
+        if self.is_done:
+            return False
+
+        container_statuses = self._pod.get("status", {}).get("container_statuses", [])
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_job.py:570]: we can add a quick comment here re: the motivation for this change

```
# Container status updates before pod phase, often by multiple seconds.
        container_statuses = self._pod.get("status", {}).get("container_statuses", [])
        if not container_statuses or len(container_statuses) < 1:
            # Yet to start, most likely
            return self._pod.get("status", {}).get("phase") == "Running"

        for cstatus in container_statuses:
            # If the terminated field is not set, the pod is still running.
            terminated = cstatus.get("state", {}).get("terminated", {})
            if not terminated or not terminated.get("finished_at"):
                # If the terminated field is set but the `finished_at` field is not 
                # set, the pod is still considered as running.
                return True
```

File: metaflow/plugins/kubernetes/kubernetes_job.py [hunk 5]
```diff
@@ -563,9 +564,29 @@ def has_failed(self):
     @property
     def is_running(self):
         # Returns true if the pod is running.
-        return not self.is_done and (
-            self._pod.get("status", {}).get("phase") == "Running"
-        )
+        if self.is_done:
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_job.py:567]: the `is_done`, `has_failed` and `has_succeeded` methods need a corresponding update as well, otherwise they will be out of sync and communicate wrong information about the state of the object.

File: metaflow/plugins/kubernetes/kubernetes_job.py (hunk 1 above)
Comment (savingoyal) [reply metaflow/plugins/kubernetes/kubernetes_job.py:342]: `atexit` was also supposed to handle [this scenario](https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/kubernetes/kubernetes_cli.py#L239) - which would break after this change.

**Review by savingoyal — CHANGES_REQUESTED** [review]
Few comments. Let's ship this as soon as these are addressed.

File: metaflow/plugins/kubernetes/kubernetes_job.py [hunk 6]
```diff
@@ -544,28 +539,128 @@ def status(self):
...
+                return False
+
+            # If finished_at is set AND reason is Completed
+            if terminated.get("reason", "").lower() != "completed":
+                return False
+
+        return True
+
+    @property
+    def have_pod_containers_succeeded(self):
```
Comment (saikonen) [inline metaflow/plugins/kubernetes/kubernetes_job.py:580]: unused property, duplicate logic to `has_container_succeeded`

**Review by saikonen — CHANGES_REQUESTED** [review]
